Python/specialize.c

```diff
@@ -2372,14 +2372,159 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
+
+static binaryopactionfunc long_float_actions[NB_OPARG_LAST+1] = {
+    [NB_ADD] = long_float_add,
+    [NB_SUBTRACT] = long_float_subtract,
+    [NB_TRUE_DIVIDE] = long_float_true_div,
+    [NB_INPLACE_MULTIPLY] = long_float_multiply,
+};
+
+static int
+binary_op_extended_specialization(PyObject *lhs, PyObject *rhs, int oparg,
+                                  binaryopguardfunc *guard, binaryopactionfunc *action)
+{
+    if (long_float_actions[oparg]) {
+        if (long_float_guard(lhs, rhs)) {
+            *guard = long_float_guard;
+            *action = long_float_actions[oparg];
+            return 1;
+        }
+    }
+    if (float_long_actions[oparg]) {
+        if (float_long_guard(lhs, rhs)) {
+            *guard = float_long_guard;
+            *action = float_long_actions[oparg];
+            return 1;
+        }
+    }
+    return 0;
+}
+
+static PyBinaryOpSpecializationDescr*
+new_binary_op_specialization_descr(binaryopguardfunc guard, binaryopactionfunc action)
+{
+    PyBinaryOpSpecializationDescr *new_descr = PyMem_Malloc(sizeof(PyBinaryOpSpecializationDescr));
+    if (new_descr == NULL) {
+        return NULL;
+    }
+    new_descr->guard = guard;
+    new_descr->action = action;
+
+    PyThreadState *tstate = _PyThreadState_GET();
+    PyBinaryOpSpecializationDescr *head = tstate->interp->binary_op_specialization_list;
+    if (head != NULL) {
+        head->prev = new_descr;
+    }
+    new_descr->next = head;
+    new_descr->prev = NULL;
+    tstate->interp->binary_op_specialization_list = new_descr;
+
+    return new_descr;
+}
+
+static void
+free_binary_op_specialization_descr(PyBinaryOpSpecializationDescr* descr)
+{
+    if (descr->prev != NULL) {
+        descr->prev->next = descr->next;
+    }
+    else {
+        PyThreadState *tstate = _PyThreadState_GET();
+        assert(tstate->interp->binary_op_specialization_list == descr);
+        tstate->interp->binary_op_specialization_list = descr->next;
+    }
+    if (descr->next != NULL) {
+        descr->next->prev = descr->prev;
+    }
+    PyMem_Free(descr);
+}
+
 void
+_Py_Specialize_FreeAllSpecializationDescrs(PyInterpreterState *interp)
+{
+    while(interp->binary_op_specialization_list) {
+        free_binary_op_specialization_descr(
+            interp->binary_op_specialization_list);
+    }
+}
+
+int
 _Py_Specialize_BinaryOp(_PyStackRef lhs_st, _PyStackRef rhs_st, _Py_CODEUNIT *instr,
                         int oparg, _PyStackRef *locals)
 {
     PyObject *lhs = PyStackRef_AsPyObjectBorrow(lhs_st);
     PyObject *rhs = PyStackRef_AsPyObjectBorrow(rhs_st);
     assert(ENABLE_SPECIALIZATION_FT);
     assert(_PyOpcode_Caches[BINARY_OP] == INLINE_CACHE_ENTRIES_BINARY_OP);
+
+    _PyBinaryOpCache *cache = (_PyBinaryOpCache *)(instr + 1);
+    if (instr->op.code == BINARY_OP_EXTEND) {
+        void *data = read_void(cache->external_cache);
+        free_binary_op_specialization_descr(data);
+        write_void(cache->external_cache, NULL);
```

Comment:
```suggestion
        write_void(cache->external_cache, NULL);
        free_binary_op_specialization_descr(data);
```
Maybe it does not matter at this moment, but it feels cleaner to first set the entry to null and then free the data.